### DISCUSSION TITLE
How to migrate from WithWillMessage to WithWillTopic?

### DISCUSSION THREAD

**Discussion by ramonsmits** [dotnet/MQTTnet#1418] [Q&A]
v4 removed WithWillMessage but WithWillTopic does not allow to pass a payload. How to do that?

**Comment by chkr1011** [comment]
In version 4 all possible values for a will message are separate properties and "With..." calls. The reason is that a will message is not a 100% compatible message which usually gets published. Some values are not available so in version 4 this fact is represented by splitting the will message into separate properties and the payload can be added with _WithWillPayload_.

**Reply by ymdeveloper** [reply]
how can I set the Retain value in v4 for the LastWillMessage ?

**Reply by chkr1011** [reply]
Unfortunately these methods are missing in the current version. I will release version 4.0.1 this weekend which adds these methods. In the meantime you can manually set the property "WillRetain" at the options after building them with the builder.
